#' update library annotation
#'
#' Check all geneIDs in library annotation file against GeneBank and get up-to-date information.
#'
#' Since information in data bases can change it is prudent
#' to refresh the library annotation from time to time.
#' This function takes every geneID in an annotation file and checks its current status
#' in GeneBank: whether is was withdrawn or replaced (and if so, by what new geneID)
#' Once the geneIDs are updated, two new queries are sent to GeneBank.
#' The first retrieves the current gene symbol, gene description, map location,
#' chromosome number, and aliases of all geneIDs.
#' The other query retrieves the gene type (protein-coding, pseudogene, etc.).
#'
#' Old geneIDs and gene symbols are kept in separate columsn, e.g. \code{old_geneid}.
#'
#' @section File format:
#' As of version 2.4.0 the function has undergone some generalization.
#' It now serves not only the original Dharmacon file but also other text files.
#' The input file may be tab- or comma delimited.
#' It must contain the following information:
#' plate number, well/position (e.g. A01), geneID and gene symbol.
#' All other columns are immaterial but will not be dropped.
#'
#' Annotation files can be updated again. In that case the columns \code{old_geneid}
#' and \code{old_gene_symbol} will remain as is and the update will be run with old geneIDs
#' rather than the updated ones.
#'
#' @section Dependencies:
#' GeneBank queries are handled with the package \code{reutils}.
#' Data is loaded (and saved) with package \code{data.table}.
#' Data processing is done with the (minimal) use of \code{dplyr} and \code{tidyr}.
#' Four internal functions are called here, see \code{Functions}.
#'
#' @section Processing time:
#' \code{check_geneid_status}
#' queries GeneBank one geneID at a time, which may appear
#' as a DDOS attack if there are too many geneIDs to check.
#' Hence, a 1 second pause is introduced before every query.
#'
#' @param infile file containing the original annotation;
#'               must be compatible with \code{[data.table]{fread}};
#'               deafults to internally sotred Dharmacon annotation from 16th May 2015
#' @param outfile (optional) path to a file to save the updated annotation
#' @param verbose print function progeress as messages or not
#'
#' @return The function either invisibly returns the updated annotation
#'         or saves to a specified path and returns nothing.
#'
#' @export
#'

update_annotation <- function(infile, outfile, verbose = FALSE) {
  if (!missing(infile) && !file.exists(infile)) stop('"infile" not found')
  # check if outfile is valid - but how? if it's the same as infile it could be overwritten...

  # load original annotation
  if (verbose) message('loading annotation')
  if (missing(infile)) infile <- 'extdata/ANNOTATION.LIBRARY.GENOMIC_20150416.original.txt'
  annotation_original <- data.table::fread(file = infile, check.names = TRUE)
  # data.table class will cause problems later on
  annotation_original <- as.data.frame(annotation_original)

  # change column names to lower case
  nms <- tolower(names(annotation_original))

  # these columns are added during the update
  # if they are all here already, it means the file was previously updated
  added_columns <- c('gene_type', 'withdrawn', 'replaced', 'gene_symbol',
                     'description', 'map_location', 'chromosome', 'aliases',
                     'old_gene_symbol', 'old_geneid')

  if (all(is.element(added_columns, nms))) {
    ## if the file has undergone a previous update
    # hold on to "old geneids" and "old gene symbols"
    hold_geneid <- annotation_original$old_geneid
    hold_gene_symbol <- annotation_original$old_gene_symbol
    # remove all added variables
    ind <- !is.element(nms, added_columns)
    annotation_original <- annotation_original[ind]
    # restore old geneids and gene symbols
    annotation_original$geneid <- hold_geneid
    annotation_original$genesymbol <- hold_gene_symbol
  } else {
    ## if this is a new file, the format must be unified
    # define a function
    replacer <- function(x, string, replacement, error, error2) {
      ind <- grep(string, x)
      if (length(ind) == 1) x[ind] <- replacement else
        if (length(ind) == 0) stop(error, call. = FALSE) else
          stop(error2, call. = FALSE)
      return(x)
    }
    # create vectors of regexs, replacements, and error messages
    strings <- c('^wells?|^positions?', 'gene[-,_,\\., ]?symbols?', 'gene[-,_,\\., ]?ids?')
    replacements <- c('position', 'genesymbol', 'geneid')
    errors <- paste('"data" contains no apparent specification of',
                    c('position', 'gene symbols', 'gene ids'))
    errors2 <- paste('"data" contains ambiguous specification of',
                     c('position', 'gene symbols', 'gene ids'))
    # run the function across the column names
    for (i in seq_along(strings)) {
      nms <- replacer(nms, strings[i], replacements[i], errors[i], errors2[i])
    }
    # replace column names
    names(annotation_original) <- nms
    ### required column names are now present
  }

  # change geneids to character, required for merging later
  annotation_original$geneid <- as.character(annotation_original$geneid)
  # extract original geneIDs
  suppressWarnings({
    geneids <- as.numeric(unique(annotation_original$geneid))
    geneids <- geneids[!is.na(geneids)]
  })

  # check geneID status
  if (verbose) message('checking geneID status')
  if (verbose) double_check <- check_geneids(geneids) else
    suppressMessages(double_check <- check_geneids(geneids))
  # add geneID status to annotation and amend geneIDs
  annotation_checked <- merge(annotation_original, double_check, by = 'geneid', all = TRUE)
  annotation_checked$old_geneid <- annotation_checked$geneid
  annotation_checked$geneid <- ifelse(is.na(annotation_checked$new_geneid),
                                      annotation_checked$old_geneid,
                                      annotation_checked$new_geneid)
  # extract new geneIDs
  suppressWarnings({
    geneids <- as.numeric(unique(annotation_checked$geneid))
    geneids <- geneids[!is.na(geneids)]
  })
  # get gene fields for all new geneIDs
  if (verbose) message('getting locus info...')
  fields <- get_gene_fields_batch(geneids)
  # append fields to amended annotation
  if (verbose) message('appending to annotation...')
  annotations_merged <- merge(annotation_checked, fields, by = 'geneid')
  # update annotation
  if (verbose) message('updating annotation...')
  # update names
  annotation_updated <- annotations_merged
  nms <- names(annotation_updated)
  nms[grepl('gene\\.?symbol', nms)] <- 'old_gene_symbol'
  nms[grepl('duplex.?catalog.?number', nms)] <- 'duplex_catalog_number'
  nms[grepl('pool.?catalog.?number', nms)] <- 'pool_catalog_number'
  nms[grepl('gene.?accession', nms)] <- 'gene_accession'
  names(annotation_updated) <- nms
  # update "plate" and "position" columns
  annotation_updated$plate <- as.numeric(gsub('[P,p]late[_,-,\\., ]?', '', annotation_updated$plate))
  annotation_updated$position <- toupper(annotation_updated$position)
  # for each plate/position wrap contents of "sequence" into single strings
  if (is.element('sequence', nms)) {
    f <- list(annotation_updated$plate, annotation_updated$position)
    sequences <- tapply(annotation_updated$sequence, f, FUN = paste, collapse = ', ')
    sequences <- unsplit(sequences, f)
    annotation_updated$sequences <- sequences
    ind <- !is.element(nms, 'sequence')
    annotation_updated <- annotation_updated[ind]
  }
  # the same for "duplex_catalog_number"
  if (is.element('duplex_catalog_number', nms)) {
    f <- list(annotation_updated$plate, annotation_updated$position)
    duplex_catalog_numbers <-
      tapply(annotation_updated$duplex_catalog_number, f, FUN = paste, collapse = ', ')
    duplex_catalog_numbers <- unsplit(duplex_catalog_numbers, f)
    annotation_updated$duplex_catalog_numbers <- duplex_catalog_numbers
    ind <- !is.element(nms, 'duplex_catalog_number')
    annotation_updated <- annotation_updated[ind]
  }
  # clean up duplicated rows, rearrange columns and rows
  annotation_updated <- annotation_updated[-which(names(annotation_updated) == 'new_geneid')]
  annotation_updated <- dplyr::select(annotation_updated, dplyr::one_of(
    c('plate', 'position', 'geneid', 'ginumber', 'gene_accession', 'gene_symbol',
      'aliases', 'description', 'map_location', 'chromosome', 'gene_type',
      'sequences', 'duplex_catalog_numbers', 'pool_catalog_number',
      'withdrawn', 'replaced', 'old_geneid', 'old_gene_symbol')),
    dplyr::everything())
  annotation_updated <- annotation_updated[!duplicated(annotation_updated), ]
  annotation_updated <- annotation_updated[order(annotation_updated$plate, annotation_updated$position), ]
  rownames(annotation_updated) <- 1:nrow(annotation_updated)

  # fill in "none"s and "???"s in geneID, gene symbol and description columns
  annotation_updated$gene_symbol <-
    ifelse(annotation_updated$geneid == 'none', 'none', annotation_updated$gene_symbol)
  annotation_updated$gene_symbol <-
    ifelse(annotation_updated$geneid == '???', '???', annotation_updated$gene_symbol)
  annotation_updated$description <-
    ifelse(annotation_updated$geneid == 'none', 'none', annotation_updated$description)
  # save or return result
  if (missing(outfile)) {
    if (verbose) message('finished')
    invisible(annotation_updated)
  } else {
    if (verbose) message('saving file')
    data.table::fwrite(annotation_updated, file = outfile, sep = '\t')
    if (verbose) message('done!')
  }
}

#' @describeIn update_annotation
#' checks a single geneID and returns its gene type (protein-coding, pseudo, etc.),
#' withdrawn and replaced status (TRUE/FALSE), and a potential new geneID;
#' returns a character vector
#'
#' @keywords internal
#'

check_geneid_status <- function(geneID) {
  if (length(geneID) != 1L) stop('"geneID" must be of length 1')
  efetch_object <- reutils::efetch(geneID, db = 'gene', rettype = 'gene_table')
  efetch_text <- reutils::content(efetch_object, as = 'text')
  withdrawn <- as.character(grepl('withdrawn', efetch_text))
  replaced <-
    if (grepl('replaced', efetch_text)) {
      grep('replaced', unlist(strsplit(efetch_text, split = '\n')), value = TRUE)
      } else NA_character_
  return(c(geneid = as.character(geneID), withdrawn = withdrawn, replaced = replaced))
}

#' @describeIn update_annotation
#' runs \code{check_geneid_status} for all geneIDs and returns a \code{data.frame};
#' pauses for 1 second before each request to avoid suspicion of DDOS attack
#'
#' @keywords internal
#'
check_geneids <- function(geneIDs) {
  # define function that counts iterations
  how_to_count <- function() {
    X <- length(geneIDs)
    iteration <- 1
    function() {
      message('\t geneID ', iteration, ' of ', X, '...')
      iteration <<- iteration + 1
    }
  }
  count <- how_to_count()
  # define checkingcounting function operator for checking geneid status
  check_geneid_status_with_pause <- function(x) {
    count()
    Sys.sleep(1)
    check_geneid_status(x)
  }
  m <- vapply(geneIDs, check_geneid_status_with_pause, character(3))
  d <- as.data.frame(t(m), stringsAsFactors = FALSE)
  d$withdrawn <- as.logical(d$withdrawn)
  d <- tidyr::separate(d, 'replaced', c('replaced', 'new_geneid'), sep = 'ID: ')
  d$replaced <- ifelse(is.na(d$replaced), FALSE, TRUE)
  d$new_geneid <- as.numeric(d$new_geneid)
  return(d)
}

#' @describeIn update_annotation
#' queries the gene data base twice;
#' first retrieves five fields:
#' gene symbol, gene description, map location, chromosome number, and aliases
#' (other geneIDs associated with the geneID);
#' then retrieves gene type
#'
#' @keywords internal
#'
get_gene_fields <- function(geneIDs) {
  # get gene symbol, aliases, description, map location and chromosome
  e_object <- reutils::efetch(geneIDs, db = 'gene', rettype = 'docsum', retmode = 'text')
  e_object_as_text <- reutils::content(e_object, as = 'text')
  split_text <- strsplit(e_object_as_text, split='<[/]?Name>')[[1]]
  gene_symbol <- split_text[seq(from = 2, to = length(split_text), by = 2)]
  split_text <- strsplit(e_object_as_text, split='<[/]?OtherAliases>')[[1]]
  aliases <- split_text[seq(from = 2, to = length(split_text), by = 2)]
  split_text <- strsplit(e_object_as_text, split='<[/]?Description>')[[1]]
  description <- split_text[seq(from = 2, to = length(split_text), by = 2)]
  split_text <- strsplit(e_object_as_text, split='<[/]?MapLocation>')[[1]]
  map_location <- split_text[seq(from = 2, to = length(split_text), by = 2)]
  split_text <- strsplit(e_object_as_text, split='<[/]?Chromosome>')[[1]]
  chromosome <- split_text[seq(from = 2, to = length(split_text), by = 2)]
  # an unexpected pause happens here; an intentional pause seems to alleviate it
  Sys.sleep(1)
  # get gene type
  efetch_object <- reutils::efetch(geneIDs, db = 'gene')
  efetch_text <- reutils::content(efetch_object, as = 'text')
  efetch_text_split <- strsplit(efetch_text, '\n')[[1]]
  gene_type_fields <- grep('gene_type', efetch_text_split, value = TRUE)
  gene_type_fields_split <- strsplit(gene_type_fields, '\"')
  ef <- function(x) {
    tryCatch(x[2], error = function(e) return('unknown gene type'))
    if (is.na(x[2])) return('unknown gene type') else return(x[2])
  }
  gene_type <- vapply(gene_type_fields_split, ef, character(1))

  return(
    data.frame(
      geneid = as.character(geneIDs), gene_symbol, aliases, description, map_location, chromosome,
      gene_type,
      stringsAsFactors = F))
}

#' @describeIn update_annotation
#' runs \code{get_gene_fields} in batches of 499 and less; this is necessary as
#' the results of \code{reutils::efetch} are unworkable for larger sets
#'
#' @keywords internal
#'
get_gene_fields_batch <- function(geneIDs) {
  # we shall be calling efetch, which can only be done for less than 500 geneIDs at a time
  # test how many items there are
  howmany <- length(geneIDs)
  # if there area less that 500 items, a single call suffices
  if (howmany < 500) {
    #if (verbose) message('... in one batch') # scoping not working...
    return(get_gene_fields(geneIDs))
  } else {
    # if there are 500 or more, we shall do it in 499-item steps
    # how many steps will there be?
    steps <- ceiling(howmany / 499)
    #if (verbose) message('... in ', steps, ' batches') # scoping not working...
    # create funcion factory that will select a 499-long intervals from a long vector
    stepper <- function(step) {
      # return function that returns the i-th 499-element section of x
      function(x) {
        X <- x[1:499 + 499 * (step -1)]
        X[!is.na(X)]
      }
    }
    # run the stepper function factory:
    # generate a list of functions that each returns a section of a vector
    steppers <- lapply(1:steps, stepper)
    # separate geneIDs into list of intervals
    ranges <- lapply(steppers, function(f) f(geneIDs))
    # do the deed over the list
    parts <- lapply(ranges, get_gene_fields)
    # wrap into single data frame
    return(do.call(rbind, parts))
  }
}


# ###
# this is a secrt stash with code left over from unifying annotations
# library(magrittr)
# library(data.table)
# library(dplyr)
#
# f <- list.files(path = '../../', pattern = '2019100[8,9]') %>% paste0('../../', .)
# l <- lapply(f, fread)
#
# ch <- function(x) {
#   if (is.element('geneid', names(x))) x$geneid <- as.character(x$geneid)
#   if (is.element('ginumber', names(x))) x$ginumber <- as.character(x$ginumber)
#   if (is.element('old_geneid', names(x))) x$old_geneid <- as.character(x$old_geneid)
#   if (is.element('chromosome', names(x))) x$chromosome <- as.character(x$chromosome)
#   return(x)
# }
#
# lc <- lapply(l, ch)
# L <- do.call(bind_rows, lc)
# glimpse(L)
# ###
